hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/JDBCSource.java

```diff
@@ -0,0 +1,233 @@
+          for (String option : options) {
+            if (!StringUtils.isNullOrEmpty(option)) {
+              String[] kv = option.split("=");
+              dataFrameReader = dataFrameReader.option(kv[0], kv[1]);
+              LOG.info("{} = {} has been set for JDBC pull ", kv[0], kv[1]);
+            }
+          }
+        }
+      }
+      if (properties.getBoolean(Config.IS_INCREMENTAL)) {
+        DataSourceUtils.checkRequiredProperties(properties, Arrays.asList(Config.INCREMENTAL_COLUMN));
+      }
+      return dataFrameReader;
+    } catch (Exception e) {
+      throw new HoodieException(e);
+    } finally {
+      IOUtils.closeStream(passwordFileStream);
+    }
+  }
+
+  @Override
+  protected Pair<Option<Dataset<Row>>, String> fetchNextBatch(Option<String> lastCkptStr, long sourceLimit) {
+    try {
+      DataSourceUtils.checkRequiredProperties(props, Arrays.asList(Config.URL, Config.DRIVER_CLASS, Config.USER,
+          Config.RDBMS_TABLE_NAME, Config.IS_INCREMENTAL));
+
+      Option<String> beginInstant =
+          lastCkptStr.isPresent() ? lastCkptStr.get().isEmpty() ? Option.empty() : lastCkptStr : Option.empty();
+      boolean isIncremental = props.getBoolean(Config.IS_INCREMENTAL);
+      if (beginInstant.equals(Option.empty())) {
+        LOG.info("No previous checkpoints found.. ");
+        Dataset<Row> rowDataset = validatePropsAndGetDataFrameReader(sparkSession, props).load();
+        return sendDfAndCheckpoint(rowDataset, isIncremental);
+      } else {
+        if (StringUtils.isNullOrEmpty(beginInstant.get())) {
+          LOG.warn("Previous checkpoint entry was null or empty. Falling back to full jdbc pull.");
+          Dataset<Row> rowDataset = validatePropsAndGetDataFrameReader(sparkSession, props).load();
+          return sendDfAndCheckpoint(rowDataset, isIncremental);
+        } else {
+          String query = String
+              .format(ppdQuery, props.getString(Config.RDBMS_TABLE_NAME), props.getString(Config.INCREMENTAL_COLUMN),
+                  beginInstant.get());
+          LOG.info("Referenced last checkpoint and prepared new predicate pushdown query for jdbc pull {}", query);
+          Dataset<Row> rowDataset = validatePropsAndGetDataFrameReader(sparkSession, props)
+              .option(props.getString(Config.RDBMS_TABLE_PROP), query).load();
+          return sendDfAndCheckpoint(rowDataset, isIncremental);
+
+        }
+      }
+    } catch (Exception e) {
+      LOG.error("Exception while running JDBCSource ", e);
+      return Pair.of(Option.empty(), null);
+    }
+  }
+
+  @NotNull
+  private Pair<Option<Dataset<Row>>, String> sendDfAndCheckpoint(Dataset<Row> rowDataset, boolean isIncremental) {
+    if (isIncremental) {
+      Column incrementalColumn = rowDataset.col(props.getString(Config.INCREMENTAL_COLUMN));
+      final String max = rowDataset.agg(functions.max(incrementalColumn).cast(DataTypes.StringType)).first()
```

Comment:
this might fetch the data out twice? once to compute the aggregation and once to actually read the values out? can we either an accumulator (if possible) or just persist the dataset before you call `.agg` ?

Reply:
As far as my knowledge it should not pull the rdd twice. However, I will still get an explain plan and check if I can see that.

Reply:
It will unless you cache it.. I am pretty certain :) // let me know what you find.